aws/aws-cloud-map-mcs-controller-for-k8s
# Propagate ClusterID to Cloud Map & Endpoints

*Affiliated with [Issue 47](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/47)*

*TLDR:*
ClusterID getting exported to Cloud Map & getting propagated to endpoints as an `CLUSTER_ID` attribute

*Description of changes:*
- Added the ClusterProperty CRD (from new group `about.k8s.io`)
- In `serviceexport_controller.go`, we now watch for changes in the ClusterProperty attribute, and when there is a change, we reconcile all endpoints. This works for adding and removing of a ClusterProperty object. 
    - This can be tested by `kubectl apply -f config/samples/about_v1alpha1_clusterproperty.yaml`
- [Update 1] If no ClusterId is defined - an error will now be thrown. Service Export will not be created.
- [Update 2 - b35bc75] ClustersetId added. This id is sent in DiscoverInstances as a QueryParameter to only return endpoints within the clusterset. More unit tests added.
- [Update 3] ClusterUtils added. **Integration tests will fail until next PR.**

*Notes:*
- This is a minimal PR. Main goal was to add the new about api. ~~No error handling introduced yet and no errors will be thrown yet for no ClusterID being defined other than when the service is being exported. No unit tests in this PR either.~~
- Adding / removing the ClusterProperty (id.k8s.io) will add/remove the `CLUSTER_ID` attribute in the endpoint
- In `suite_test.go`, changes needed to be made to not include the clusterproperty CRD for testing. Special tests will need to be constructed to manually add the annotation to be allowed to use '.k8s.io' suffix.

*Testing:*
- manual testing of the controller
- `make test` runs without error
- [Update 1] tests updated to include ClusterID. Unit test added for no cluster id - expect
- [Update 2] more unit tests added for Clusterid and ClustersetId
- [Update 3] Integration tests will fail until next PR
